src/array-map.ts

```diff
@@ -0,0 +1,293 @@
+   *    map: (num) => `hi, ${num}!`
+   *  });
+   *  // ...
+   * ```
+   * ```hbs
+   *  {{#each this.myMappedData as |datum|}}
+   *     loop body only invoked for changed entries
+   *     {{datum}}
+   *  {{/each}}
+   * ```
+   *
+   * Iteration in javascript is also provided by this iterator
+   * ```js
+   *  class Foo {
+   *    myMappedData = map(this, {
+   *      data: () => [1, 2, 3],
+   *      map: (num) => `hi, ${num}!`
+   *    });
+   *
+   *    get mapAgain() {
+   *      let results = [];
+   *
+   *      for (let datum of this.myMappedData) {
+   *        results.push(datum);
+   *      }
+   *
+   *      return datum;
+   *    }
+   *  }
+   * ```
+   */
+  [Symbol.iterator](): Iterator<MappedTo>;
+}
+
+/**
+ * Reactivily apply a `map` function to each element in an array,
+ * persisting map-results for each object, based on identity.
+ *
+ * This is useful when you have a large collection of items that
+ * need to be transformed into a different shape (adding/removing/modifying data/properties)
+ * and you want the transform to be efficient when iterating over that data.
+ *
+ * A common use case where this `map` utility provides benefits over is
+ * ```js
+ * class MyClass {\
+ *   @signal
+ *   get wrappedRecords() {
+ *     return this.records.map(record => new SomeWrapper(record));
+ *   }
+ * }
+ * ```
+ *
+ * Even though the above is a cached computed (via `@signal`), if any signal data accessed during the evaluation of `wrappedRecords`
+ * changes, the entire array.map will re-run, often doing duplicate work for every unchanged item in the array.
+ *
+ * @return {MappedArray} an object that behaves like an array. This shouldn't be modified directly. Instead, you can freely modify the data returned by the `data` function, which should be auto-tracked in order to benefit from this abstraction.
+ *
+ * @example
+ *
+ * ```js
+ *  import { arrayMap } from 'signal-utils/array-map';
+ *
+ *  class MyClass {
+ *    wrappedRecords = map({
+ *      data: () => this.records,
+ *      map: (record) => new SomeWrapper(record),
+ *    }),
+ *  }
+ * ```
+ */
+export function arrayMap<
+  Elements extends readonly unknown[],
+  MapTo = unknown,
+>(options: {
+  /**
+   * Array of non-primitives to map over
+   *
+   * This can be class instances, plain objects, or anything supported by WeakMap's key
+   */
+  data: () => Elements;
+  /**
+   * How to transform each element from `data`,
+   * similar to if you were to use Array map yourself.
+   *
+   * This function will be called only when needed / on-demand / lazily.
+   * - if iterating over part of the data, map will only be called for the elements observed
+   * - if not iterating, map will only be called for the elements observed.
+   */
+  map: (element: Elements[0]) => MapTo;
+}): MappedArray<Elements, MapTo> {
+  let { data, map } = options;
+
+  return new TrackedArrayMap(data, map) as MappedArray<Elements, MapTo>;
+}
+
+const AT = Symbol.for("__AT__");
```

Comment:
To avoid collisions with other uses of `__AT__` should this not be a newly-created (unique) symbol rather than `Symbol.for`?

Reply:
yea, I was originally using the collision on purpose in tests, but I've reverted all that. This will go back to being a unique symbol. thanks!